cpp/src/arrow/compute/expression.cc

```diff
@@ -845,7 +845,7 @@ Result<Expression> FoldConstants(Expression expr) {
       std::move(expr), [](Expression expr) { return expr; },
       [](Expression expr, ...) -> Result<Expression> {
         auto call = CallNotNull(expr);
-        if (call->function->is_impure()) return expr;
+        if (!call->function->is_pure()) return expr;
```

Comment:
The syntax seems to be more intuitive. The previous modification considered that the call of impure expression requires special processing, and only the impure itself is used.

Reply:
Understandable. You did what is best for what you were trying to accomplish — skip impure functions, but I'm trying to simplify the configuration/implementation for function implementers.